### ISSUE TITLE
Feature request: implement fish shell like functions folder in ~/.elvish

### ISSUE THREAD

**Issue by sandeep-datta** [elves/elvish#1164]
Hi, please consider implementing a fish-shell like `functions` folder in `~/.elvish` or `~/.config/elvish`. Any `.elv` files in that folder should be automatically loaded at startup. This will let the user create aliases and reusable functions quickly.

**Comment by hanche** [comment]
I think it is an interesting proposal, but it raises a number of concerns.

1. I suggest the file should be loaded in a namespace of its own. That way, it can define private helper functions too, and variables maintaining state.
2. Requiring the file to define a function with the same name as the file makes it harder to rename the function, as you both have to rename the file and edit it. An alternative is either to use a fixed name, such as `main`, or to have the file `put` the defined function. In either case, if the protocol is not followed, that should be detected and raise an error.
3. We probably want to cache the function, so the file is not loaded twice. One way would be to have a magical namespace where the function names are stored, and which is searched between the current namespace and the builtin namespace, say. To gain a bit of flexibility, allowing the user to change the file and run the modified function, the magical namespace could be made explicitly available at the elvish level. It could support the `del` operation, but otherwise be read only in order to avoid abuse. As an alternative, elvish could check the timestamp of the file and reload it if it has changed.

**Comment by krader1961** [comment]
See http://fishshell.com/docs/current/index.html#autoloading-functions for more information about the fish shell feature. Note that fish autoloads functions in both interactive and non-interactive shells. It might make more sense for this mechanism to only apply to interactive context in Elvish. If a module or function (even one defined interactively) wants to use an autoloaded function it might make more sense to require an explicit `use`; e.g., `use ff; ff:main` to use my "find files" function.

Presumably we would implement this via an implicit `use` (which [creates a namespace](https://www.skepticism.us/elvish/ref/language.html#modules)) coupled with inserting an entry in the `$-exports-` map (see #1110). I like @hanche's idea of requiring a `main` function (or a similar name) in the autoloaded module rather than a function with the same name as the module (minus the *.elv* extension).

**Comment by hanche** [comment]
> It might make more sense for this mechanism to only apply to interactive context in Elvish.

I think that could be very confusing to the user. Consider:
```
foo bar # works, by loading foo.elv
if ?(something) { foo bar } # doesn't work
```

**Comment by sandeep-datta** [comment]
>   I suggest the file should be loaded in a namespace of its own. That way, it can define private helper functions too, and variables maintaining state.

Will the user have to access the command as `namespace:user_function`? I'd prefer a mode/keyword where `use` lets the user omit the namespace for good. I think one of the distinguishing characteristics of a shell scripting language is its focus on less typing (helps avoid RSI).

>  Requiring the file to define a function with the same name as the file makes it harder to rename the function, as you both have to rename the file and edit it. An alternative is either to use a fixed name, such as `main`, or to have the file `put` the defined function. In either case, if the protocol is not followed, that should be detected and raise an error.

Great idea!

>  an alternative, elvish could check the timestamp of the file and reload it if it has changed.

I think this is simpler and easier to understand.

> `use ff; ff:main`

This is triggering my RSI. What if we make modules / scripts behave more like functions? Many main stream scripting languages use this idea. Usually the script has access to the command line arguments through automatically defined variables(`$1`, `$*` etc). But I prefer to have a (single) `main` function per script with explicitly defined arguments which let's us do this: `use ff; ff arg1 arg2 ...`. Where `ff` is a module / script.

> I think that could be very confusing to the user.

Yeah, I agree.

**Comment by hanche** [comment]
> Will the user have to access the command as `namespace:user_function`?

No, the idea was that the user would only get access to the function, not to the namespace. But my proposal was that the function will exist in its own anonymous namespace, which then has room for auxiliary variables and helper functions. In essence, if the file `foo.elv` contains
```
fn a [x y]{…}
u = something
fn main [x]{ a $u $x }
```
and there is no function named `foo`, then running `foo bar` would be equivalent to running
```
foo~ = ({
  fn a [x y]{…}
  u = something
  fn main [x]{ a $u $x }
  put $main
})
foo bar
```
with the caveat that the analogy fails if other code in `foo.elv` were to write to the value pipe. In reality, instead of the `put`, elvish would inspect the anonymous namespace and look for the `main` function, then assign that.

**Comment by zzamboni** [comment]
This is more or less what my [alias](https://github.com/zzamboni/elvish-modules/blob/master/alias.org) module does - the files from `~/.elvish/aliases/` are read and returned into the `$-exports-` special map during Elvish initialization.

**Comment by sandeep-datta** [comment]
@zzamboni thanks! I like your module but it needs to be mentioned that "Once an alias is defined, it does not become immediately available". This is a serious drawback in my opinion. I presume elvish is missing some functionality for you to do away with this workaround. Have you already filed an issue to unblock this?

**Comment by zzamboni** [comment]
@sandeep-datta this is because Elvish only allows defining top-level functions in the interactive namespace from within `rc.elv` by populating the special `$-exports-` map. Honestly I don't think starting a new Elvish session is too much trouble.
